packages/entity-server/src/models/types.ts

```diff
@@ -6,12 +6,49 @@
 import { DatabaseModel, DatabaseType } from '@tupaia/database';
 import { ObjectLikeKeys, Flatten } from '../types';
 
-type PartialOrArray<T> = {
-  [field in keyof T]?: T[field] | T[field][];
+type FilterComparators = '!=' | 'ilike';
+
+type AdvancedFilterValue<T> = {
+  comparator: FilterComparators;
+  comparisonValue: T | T[];
+};
+
+/**
+ * Example:
+ * {
+ *    type: 'facility',
+ *    name: {
+ *        comparator: 'ilike',
+ *        comparisonValue: '%Pallet%'
+ *    }
+ * }
+ */
+type NormalCriteria<T> = {
+  [field in keyof T]?: T[field] | T[field][] | AdvancedFilterValue<T[field]>;
 };
 
-export type DbConditional<T> = PartialOrArray<Omit<T, ObjectLikeKeys<T>>> &
-  PartialOrArray<Flatten<Pick<T, ObjectLikeKeys<T>>, '->>'>>;
+export enum QueryConjunctions {
+  AND = '_and_',
+  OR = '_or_',
+  RAW = '_raw_',
+}
+
+/**
+ * Example:
+ * {
+ *    _and_: {
+ *        country_code: ['TO', 'KI']
+ *    }
+ * }
+ */
+type ConjunctionCriteria<T> = {
+  [field in QueryConjunctions]?: NormalCriteria<T>;
+}
+
+type FilterCriteria<T> = NormalCriteria<T> & ConjunctionCriteria<T>;
+
+export type DbConditional<T> = FilterCriteria<Omit<T, ObjectLikeKeys<T>>> &
+  FilterCriteria<Flatten<Pick<T, ObjectLikeKeys<T>>, '->>'>>;
```

Comment:
Have to admit I'm not very confident with these types. I'm not also sure if they should sit in this file. I'm putting them in this file because they are closely related to the BaseModel

Reply:
You've actually done such an amazing job with these types! I know they're super hard to wrap your head around but I think you've got it almost perfect!